### DISCUSSION TITLE
How to use mmdeploy::Span container？

### DISCUSSION THREAD

**Discussion by Ruid6** [open-mmlab/mmdeploy#2235] [Q&A]
I want batch inference，use this function . it not work, **Can anyone give me some examples?**
![7b866dc0e97f2df706c497792ba914a](https://github.com/open-mmlab/mmdeploy/assets/48819558/351e88b4-e141-4a6a-9d01-ece192f6c172)
i dont know how to use, i use it like this

>   mmdeploy::Span \<mmdeploy::PoseTracker::State> states(&state,3);
>   mmdeploy::Span \<int32_t> detects{ -1,-1,-1 };
>   mmdeploy::Span \<const cv::Mat>  inputs{ &frame0,&frame1 };

it output error,

> 1>main.cxx
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\pose_tracker.hpp(127): warning C4267: “参数”: 从“size_t”转换到“int”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\common.hpp(64,1): message : 查看对正在编译的函数 模板 实例化“auto mmdeploy::cxx::PoseTracker::Apply::<lambda_1>::operator ()<std::nullptr_t>(_T1) const”的引用
> 1>        with
> 1>        [
> 1>            _T1=std::nullptr_t
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\pose_tracker.hpp(127): message : 查看指向正在编译的变量模板“const bool conjunction_v<std::is_move_constructible<`mmdeploy::cxx::PoseTracker::Apply'::`2'::<lambda_1> >,std::_Can_call_function_object<`mmdeploy::cxx::PoseTracker::Apply'::`2'::<lambda_1> &,std::nullptr_t &,void> >”的引用
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\utils\mediaio.h(244,27): warning C4018: “>”: 有符号/无符号不匹配
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\utils\visualize.h(64,62): warning C4244: “参数”: 从“int”转换到“_Tp”，可能丢失数据
> 1>        with
> 1>        [
> 1>            _Tp=float
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\utils\visualize.h(65,51): warning C4244: “参数”: 从“_Tp”转换到“_Tp”，可能丢失数据
> 1>        with
> 1>        [
> 1>            _Tp=int
> 1>        ]
> 1>        and
> 1>        [
> 1>            _Tp=float
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\utils\visualize.h(69,66): warning C4244: “参数”: 从“_Tp”转换到“_Tp”，可能丢失数据
> 1>        with
> 1>        [
> 1>            _Tp=int
> 1>        ]
> 1>        and
> 1>        [
> 1>            _Tp=float
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\utils\visualize.h(140,76): warning C4244: “参数”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\pose_tracker_params.h(29,21): warning C4244: “=”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\pose_tracker_params.h(30,31): warning C4244: “=”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\pose_tracker_params.h(32,26): warning C4244: “=”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\pose_tracker_params.h(34,29): warning C4244: “=”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\pose_tracker_params.h(35,32): warning C4244: “=”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\pose_tracker_params.h(36,26): warning C4244: “=”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\pose_tracker_params.h(37,27): warning C4244: “=”: 从“double”转换到“float”，可能丢失数据
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(59,31): error C2440: “初始化”: 无法从“const cv::Mat *”转换为“T *”
> 1>        with
> 1>        [
> 1>            T=const mmdeploy::cxx::Mat
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(59,31): message : 指向的类型不相关; 转换需要 reinterpret_cast、C 样式强制转换或带圆括号的函数样式强制转换
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\main.cxx(75,92): message : 查看对正在编译的函数 模板 实例化“mmdeploy::Span<const mmdeploy::cxx::Mat>::Span<mmdeploy::Span<const cv::Mat>,void,void>(U &)”的引用
> 1>        with
> 1>        [
> 1>            U=mmdeploy::Span<const cv::Mat>
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\main.cxx(75,47): message : 查看对正在编译的函数 模板 实例化“mmdeploy::Span<const mmdeploy::cxx::Mat>::Span<mmdeploy::Span<const cv::Mat>,void,void>(U &)”的引用
> 1>        with
> 1>        [
> 1>            U=mmdeploy::Span<const cv::Mat>
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(59): error C2439: “mmdeploy::Span<const mmdeploy::cxx::Mat>::data_”: 未能初始化成员
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(115,6): message : 参见“mmdeploy::Span<const mmdeploy::cxx::Mat>::data_”的声明

**Comment by irexyc** [comment]
It's a implementation of std::span, the usage should be same, you can refer to https://en.cppreference.com/w/cpp/container/span

**Reply by Ruid6** [reply]
i have seen above ,but it still not work, How to init cv::Mat ?
```
void process_data(std::span<int> data) {
    // 处理 data 内存区间的元素
}

int main() {
    int arr[] = {1, 2, 3, 4, 5};
    std::vector<int> vec = {6, 7, 8, 9, 10};
    process_data(arr);              // 传递整个数组 arr
    process_data(vec);              // 传递整个向量 vec
    process_data(arr + 1, 3);       // 传递数组 arr 的第 2~4 个元素
    process_data(vec.data() + 1, 3); // 传递向量 vec 的第 2~4 个元素
    return 0;
}
```
```
std::vector<cv::Mat> inputs = {frame0.clone(),frame1.clone()};
std::vector<mmdeploy::PoseTracker::Result> results = tracker.Apply(states,inputs);
```

> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(59,31): error C2440: “初始化”: 无法从“_Ty *”转换为“T *”
> 1>        with
> 1>        [
> 1>            _Ty=cv::Mat
> 1>        ]
> 1>        and
> 1>        [
> 1>            T=const mmdeploy::cxx::Mat
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(59,31): message : 指向的类型不相关; 转换需要 reinterpret_cast、C 样式强制转换或带圆括号的函数样式强制转换
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\main.cxx(75,84): message : 查看对正在编译的函数 模板 实例化“mmdeploy::Span<const mmdeploy::cxx::Mat>::Span<std::vector<cv::Mat,std::allocator<cv::Mat>>,void,void>(U &)”的引用
> 1>        with
> 1>        [
> 1>            U=std::vector<cv::Mat,std::allocator<cv::Mat>>
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\main.cxx(75,47): message : 查看对正在编译的函数 模板 实例化“mmdeploy::Span<const mmdeploy::cxx::Mat>::Span<std::vector<cv::Mat,std::allocator<cv::Mat>>,void,void>(U &)”的引用
> 1>        with
> 1>        [
> 1>            U=std::vector<cv::Mat,std::allocator<cv::Mat>>
> 1>        ]
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(59): error C2439: “mmdeploy::Span<const mmdeploy::cxx::Mat>::data_”: 未能初始化成员
> 1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\core\mpl\span.h(115,6): message : 参见“mmdeploy::Span<const mmdeploy::cxx::Mat>::data_”的声明

**Comment by irexyc** [comment]
There is no convertion of cv::Mat* to mmdeploy::cxx::Mat*, but has a implicit convertion of cv::Mat to mmdeploy::cxx::Mat

You can use this way:
```
std::vector<cv::Mat> inputs = {frame0.clone(),frame1.clone()};
std::vector<Mat> inputs2(inputs.begin(), inputs.end());
std::vector<mmdeploy::PoseTracker::Result> results = tracker.Apply(states,inputs2);
```

**Reply by Ruid6** [reply]
I'm sorry, again. I don't know much about C++
```
  mmdeploy::PoseTracker tracker(mmdeploy::Model(ARGS_det_model), mmdeploy::Model(ARGS_pose_model),
                                mmdeploy::Device{FLAGS_device});

  mmdeploy::PoseTracker::Params params;
  // initialize tracker params with program arguments
  InitTrackerParams(params);

  // create a tracker state for each video
  mmdeploy::PoseTracker::State state = tracker.CreateState(params);
  mmdeploy::Span<mmdeploy::PoseTracker::State> states(&state, 2);
```

> 0x00007FFCB70C97E6 (mmdeploy.dll)处(位于 testpose_tracker.exe 中)引发的异常: 0xC0000005: 读取位置 0xFFFFFFFFFFFFFFFF 时发生访问冲突。
```

  mmdeploy::PoseTracker::State state = tracker.CreateState(params);
  std::vector< mmdeploy::PoseTracker::State> states = {&state,&state};
```

> 1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\xmemory(680,47): error C2280: “mmdeploy::cxx::PoseTracker::State::State(const mmdeploy::cxx::PoseTracker::State &)”: 尝试引用已删除的函数

it work in python 
```
   
    videos = []
    states = []
    detects = []
    for i in range(args.video_nums):
        videos.append(cv2.VideoCapture(i))
        states.append(state)
        detects.append(-1)
    states = np.ascontiguousarray(np.asarray(states))
    detects = np.ascontiguousarray(np.asarray(detects))
```

**Comment by irexyc** [comment]
**span is used to describe a continuous memory sequence**

mmdeploy::PoseTracker::State state1 = tracker.CreateState(params);
mmdeploy::PoseTracker::State state2 = tracker.CreateState(params);
std::vector<mmdeploy::PoseTracker::State> vec{state1, state2}
mmdeploy::Span<mmdeploy::PoseTracker::State> states(vec.data(), 2);

**Reply by Ruid6** [reply]
i try it ,too
```
  mmdeploy::PoseTracker::State state = tracker.CreateState(params);
  mmdeploy::PoseTracker::State state1 = tracker.CreateState(params);
  std::vector< mmdeploy::PoseTracker::State> vec{ state,state1 };
  mmdeploy::Span<mmdeploy::PoseTracker::State> states(vec.data(), 2);
```
![image](https://github.com/open-mmlab/mmdeploy/assets/48819558/60f0b682-25dd-4304-8e7d-552464678ed9)

```
1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\example\cpp\cpp\main.cxx(42,46): error C2280: “mmdeploy::cxx::PoseTracker::State::State(const mmdeploy::cxx::PoseTracker::State &)”: 尝试引用已删除的函数
1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\pose_tracker.hpp(100,3): message : 编译器已在此处生成“mmdeploy::cxx::PoseTracker::State::State”
1>E:\Rtmpose\mmdeploy-1.1.0-windows-amd64-cuda11.3\include\mmdeploy\pose_tracker.hpp(100,3): message : “mmdeploy::cxx::PoseTracker::State::State(const mmdeploy::cxx::PoseTracker::State &)”: 由于“mmdeploy::cxx::PoseTracker::State”具有用户定义的 移动构造函数，因此已隐式删除函数
```

**Comment by irexyc** [comment]
my bad, try this:
```
std::array<mmdeploy::PoseTracker::State, 2> arr = {tracker.CreateState(params), tracker.CreateState(params)};
mmdeploy::Span<mmdeploy::PoseTracker::State> states(arr.data(), 2); // or just mmdeploy::Span<mmdeploy::PoseTracker::State> states(arr)
```
